src/cpu_shader/euler.rs

```diff
@@ -49,13 +49,39 @@ pub struct EulerSeg {
 impl CubicParams {
     /// Compute parameters from endpoints and derivatives.
     ///
-    /// Robustness note: this function must be protected from being called when the
-    /// chord is near zero.
+    /// This function is designed to be robust across a wide range of inputs. In
+    /// particular, it splits between near-zero chord length and the happy path.
+    /// In the former case, the parameters for the Euler spiral would not be valid,
+    /// so it proposes a straight line and computes a pretty good (conservative)
+    /// estimate of the Fréchet distance between that line and the source cubic.
+    ///
+    /// Computing an accurate estimate here fixes two tricky cases: very short
+    /// lines, in which the error will be below threshold and the flatten logic will
+    /// output a single line segment without subdividing, and loop cases with a
+    /// short chord, in which case the error will exceed the threshold, and the
+    /// chords of the subdivided pieces will be longer.
+    ///
+    /// An additional case is the near-cusp where the proposed Euler spiral has
+    /// a 180 degree U-turn (or, more generally, one angle exceeds 90 degrees and
+    /// the other does not). In that case, the resulting Euler spiral is quite
+    /// well defined (with finite curvature, so that its offset will generate a
+    /// near-semicircle, preserving G1 continuity), but the analytic error
+    /// calculation would be a huge overestimate. In that case, we just return
+    /// a rough estimate of the distance between the chord and the spiral segment.
     pub fn from_points_derivs(p0: Vec2, p1: Vec2, q0: Vec2, q1: Vec2, dt: f32) -> Self {
         let chord = p1 - p0;
-        let length_squared = chord.length_squared();
-        assert_ne!(length_squared, 0.0);
-        let scale = dt / length_squared;
+        let chord_squared = chord.length_squared();
+        let chord_len = chord_squared.sqrt();
+        if chord_squared < TANGENT_THRESH.powi(2) {
```

Comment:
nit: Consider adding a comment above this if statement like `// straight line case`. It will help the reader correlate the code with the documentation above.